File: data/derived_cot/rq2_eval/translation/qwen/_converted_code/java_to_cpp/DiscountStrategy.cpp
#include <vector>
#include <string>
#include <functional>

class Customer {
public:
    Customer(const std::string& name, int fidelity) : name(name), fidelity(fidelity) {}
    std::string getName() const { return name; }
    int getFidelity() const { return fidelity; }

private:
    std::string name;
    int fidelity;
};

class Cart {
public:
    Cart() = default;
    void addProduct(const Product& product) { products.push_back(product); }
    const std::vector<Product>& getProducts() const { return products; }

private:
    std::vector<Product> products;
};

class Product {
public:
    Product(const std::string& name, int quantity, double price) : name(name), quantity(quantity), price(price) {}
    std::string getName() const { return name; }
    int getQuantity() const { return quantity; }
    double getPrice() const { return price; }

private:
    std::string name;
    int quantity;
    double price;
};

class DiscountStrategy {
public:
    using Promotion = std::function<double(const DiscountStrategy&)>;

    static const Promotion FIDELITY_PROMO;
    static const Promotion BULK_ITEM_PROMO;
    static const Promotion LARGE_ORDER_PROMO;

    DiscountStrategy(const Customer& customer, const Cart& cart, const Promotion& promotion)
        : customer(customer), cart(cart), promotion(promotion), total(0.0) {}

    double total() const;
    double due() const;
    double promotion(const DiscountStrategy& order) const;

private:
    Customer customer;
    Cart cart;
    Promotion promotion;
    mutable double total;
};

// Define static members
const DiscountStrategy::Promotion DiscountStrategy::FIDELITY_PROMO = [](const DiscountStrategy& order) {
    return order.customer.getFidelity() >= 1000 ? order.total() * 0.05 : 0;
};

const DiscountStrategy::Promotion DiscountStrategy::BULK_ITEM_PROMO = [](const DiscountStrategy& order) {
    double discount = 0.0;
    for (const auto& item : order.cart.getProducts()) {
        if (item.getQuantity() >= 20) {
            discount += item.getQuantity() * item.getPrice() * 0.1;
        }
    }
    return discount;
};

const DiscountStrategy::Promotion DiscountStrategy::LARGE_ORDER_PROMO = [](const DiscountStrategy& order) {
    return order.cart.getProducts().size() >= 10 ? order.total() * 0.07 : 0;
};

double DiscountStrategy::total() const {
    total = 0.0;
    for (const auto& product : cart.getProducts()) {
        total += product.getQuantity() * product.getPrice();
    }
    return total;
}

double DiscountStrategy::due() const {
    double discount = (promotion == nullptr) ? 0 : promotion(*this);
    return total - discount;
}

double DiscountStrategy::promotion(const DiscountStrategy& order) const {
    return (promotion == nullptr) ? 0 : promotion(order);
}

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_py/code_output/DiscountStrategy.py
class Customer:
    def __init__(self, name, fidelity):
        self.name = name
        self.fidelity = fidelity

    def getName(self):
        return self.name

    def getFidelity(self):
        return self.fidelity


class Product:
    def __init__(self, name, quantity, price):
        self.name = name
        self.quantity = quantity
        self.price = price

    def getName(self):
        return self.name

    def getQuantity(self):
        return self.quantity

    def getPrice(self):
        return self.price


class Cart:
    def __init__(self, *products):
        self.products = list(products)

    def addProduct(self, product):
        self.products.append(product)

    def getProducts(self):
        return self.products


class DiscountStrategy:
    FIDELITY_PROMO = None
    BULK_ITEM_PROMO = None
    LARGE_ORDER_PROMO = None

    def __init__(self, customer, cart, promotion):
        self.customer = customer
        self.cart = cart
        self._promotion = promotion
        self._total = self.total()

    def total(self):
        self._total = sum(p.getQuantity() * p.getPrice() for p in self.cart.getProducts())
        return self._total

    def due(self):
        discount = 0 if self._promotion is None else self._promotion(self)
        return self._total - discount

    def promotion(self, order):
        return 0 if self._promotion is None else self._promotion(self)


DiscountStrategy.FIDELITY_PROMO = lambda order: order.total() * 0.05 if order.customer.getFidelity() >= 1000 else 0

DiscountStrategy.BULK_ITEM_PROMO = lambda order: sum(
    item.getQuantity() * item.getPrice() * 0.1
    for item in order.cart.getProducts()
    if item.getQuantity() >= 20
)

DiscountStrategy.LARGE_ORDER_PROMO = lambda order: order.total() * 0.07 if len(order.cart.getProducts()) >= 10 else 0